HIVE-21052: Make sure transactions get cleaned if they are aborted be…

deniskuzZ opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://cwiki.apache.org/confluence/display/Hive/HowToContribute
  2. Ensure that you have created an issue on the Hive project JIRA: https://issues.apache.org/jira/projects/HIVE/summary
  3. Ensure you have added or run the appropriate tests for your PR: 
  4. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP]HIVE-XXXXX:  Your PR title ...'.
  5. Be sure to keep the PR description updated to reflect all changes.
  6. Please write your PR title to summarize what this PR proposes.
  7. If possible, provide a concise example to reproduce the issue for a faster review.

-->

### What changes were proposed in this pull request?

Below changes are only with respect to master.

Design: taken from https://issues.apache.org/jira/secure/attachment/12954375/Aborted%20Txn%20w_Direct%20Write.pdf

**Overview:**
1. add a dummy row to TXN_COMPONENTS with operation type 'p' in enqueueLockWithRetry, which will be removed in addDynamicPartition
2. If anytime txn is aborted, this dummy entry will be block initiator to remove this txnId from TXNS
3. Initiator will add a row in COMPACTION_QUEUE (with type 'p') for the above aborted txn with the state as READY_FOR_CLEANING, at a time there will be a single entry of this type for a table in COMPACTION_QUEUE.
4. Cleaner will directly pickup above request, and process it via new cleanAborted code path(scan all partitions and remove aborted dirs), once successful cleaner will remove dummy row from TXN_COMPONENTS

**Cleaner Design:**
- We are keeping cleaner single thread, and this new type of cleanup will be handled similar to any regular cleanup

**Aborted dirs cleanup:**
- In p-type cleanup, cleaner will iterate over all the partitions and remove all delta/base dirs with given aborted writeId list
- added cleanup of aborted base/delta in the worker also

**TXN_COMPONENTS cleanup:**
- If successful, p-type entry will be removed from TXN_COMPONENTS during addDynamicPartitions
- If aborted, cleaner will clean in markCleaned after successful processing of p-type cleanup

**TXNS cleanup:**
- No change, will be cleaned up by the initiator 

<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
To fix above mentioned issue.
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce _any_ user-facing change?
No
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description, screenshot and/or a reproducable example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Hive versions or within the unreleased branches such as master.
If no, write 'No'.
-->


### How was this patch tested?
unit-tests added
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

deniskuzZ commented:
@vpnvishv , could you please check.

vpnvishv commented on ql/src/test/org/apache/hadoop/hive/ql/TestTxnCommands2.java:
```diff
@@ -2128,6 +2128,395 @@ public void testCleanerForTxnToWriteId() throws Exception {
+    runStatementOnDriver("alter table "+ Table.MMTBL + " compact 'MINOR'");
+    // The worker should remove the subdir for aborted transaction
+    runWorker(hiveConf);
+    verifyDeltaDirAndResult(2, Table.MMTBL.toString(), "", resultData1);
```
Here we need some correction in expected values, as these tests were written wrt to branch-3.1 in which we don't support minor compaction for MM tables.

deniskuzZ replied:
fixed

vpnvishv commented on ql/src/test/org/apache/hadoop/hive/ql/TestTxnCommands2.java:
```diff
@@ -2128,6 +2128,395 @@ public void testCleanerForTxnToWriteId() throws Exception {
+    runStatementOnDriver("insert into " + Table.MMTBL + "(a,b) values(5,6)");
+    verifyDeltaDirAndResult(3, Table.MMTBL.toString(), "", resultData2);
+    r1 = runStatementOnDriver("select count(*) from " + Table.MMTBL);
+    Assert.assertEquals("2", r1.get(0));
```
Here expected value looks correct to me and this looks like a genuine issue. Is this due to some incorrect stats, any idea?

deniskuzZ replied:
fixed, turned off StatsOptimizer

vpnvishv commented on ql/src/test/org/apache/hadoop/hive/ql/TestTxnCommands2.java:
```diff
@@ -2128,6 +2128,395 @@ public void testCleanerForTxnToWriteId() throws Exception {
+    runStatementOnDriver("insert into " + Table.ACIDTBL + "(a,b) values(5,6)");
+    verifyDeltaDirAndResult(3, Table.ACIDTBL.toString(), "", resultData2);
+    r1 = runStatementOnDriver("select count(*) from " + Table.ACIDTBL);
+    Assert.assertEquals("2", r1.get(0));
```
Again, here expected value looks correct to me and this looks like a genuine issue. Is this due to some incorrect stats, any idea?

deniskuzZ replied:
fixed, turned of StatsOptimizer

vpnvishv reviewed (commented):
@deniskuzZ Overall change LGTM.
Looked into the test failures, one of test requires change in expected values wrt master branch. Other two looks genuine failures to me. Please check the inline comments.

pvargacl commented on ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:
```diff
@@ -2839,6 +2848,87 @@ public static void setNonTransactional(Map<String, String> tblProps) {
+    return deleted;
+  }
+
+  private static void deleteDeltaDirectoriesAux(Path root, FileSystem fs, PathFilter filter, List<FileStatus> deleted)
```
This is going issue many filesystem listing on a table with many partitions, that is going to be very slow on S3. I think you should consider changing this logic to be similar to getHdfsDirSnapshots that would do 1 recursive listing, iterate all the files and collect the deltas that needs to be deleted and delete them at the end (possible concurrently)

deniskuzZ replied:
getHdfsDirSnapshots does the same recursive listing, isn't it?
```
RemoteIterator<LocatedFileStatus> itr = fs.listFiles(path, true);
while (itr.hasNext()) {
```

deniskuzZ replied:
changed to use getHdfsDirSnapshots, 
@pvargacl do you know. if i should access cached data somehow?

pvargacl commented on ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:
```diff
@@ -2839,6 +2848,87 @@ public static void setNonTransactional(Map<String, String> tblProps) {
+    PathFilter filter = (p) -> {
+      String name = p.getName();
+      for (Long wId : writeIds) {
+        if (name.startsWith(deltaSubdir(wId, wId)) && !name.contains("=")) {
```
Why the contains "=", are we checking for a partition where the user named the column exactly like a valid delta dir? I don't think we should support that

vpnvishv replied:
I was also wondering the same, as this code was there in the earlier patches so I have just kept it. We can remove this.

deniskuzZ replied:
changed, also excluded base directory from listing

pvargacl commented on ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:
```diff
@@ -2839,6 +2848,87 @@ public static void setNonTransactional(Map<String, String> tblProps) {
+          deleted.add(fStatus);
+        } else {
+          deleteDeltaDirectoriesAux(fStatus.getPath(), fs, filter, deleted);
+          if (isDirectoryEmpty(fs, fStatus.getPath())) {
```
Are we doing this to delete newly created partitions if there are no other writes? Is this ok, what if we found a valid empty partition that is registered in the HMS? We should not delete that. I think this can be skipped all together, the empty partition dir will not bother anybody

deniskuzZ replied:
agree, that would simplify re-use of getHdfsDirSnapshots

deniskuzZ replied:
+ partitions are not removed in HMS

pvargacl commented on ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:
```diff
@@ -2839,6 +2848,87 @@ public static void setNonTransactional(Map<String, String> tblProps) {
+      throws IOException {
+    try {
+      return new ToFileStatusIterator(SHIMS.listLocatedHdfsStatusIterator(fs, path, filter));
+    } catch (Throwable t) {
```
This should be similar to tryListLocatedHdfsStatus don't catch all Throwable

deniskuzZ replied:
removed it

pvargacl commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:
```diff
@@ -97,9 +100,9 @@ public void run() {
+          for (CompactionInfo compactionInfo : txnHandler.findReadyToClean()) {
             cleanerList.add(CompletableFuture.runAsync(CompactorUtil.ThrowingRunnable.unchecked(() ->
-                    clean(compactionInfo, minOpenTxnId)), cleanerExecutor));
+                  clean(compactionInfo, minOpenTxnId)), cleanerExecutor));
```
Two questions here:
1. In the original Jira there was discussion about not allowing concurrent cleanings of the same stuff (partition / table). Should we worry about this?
2. The slow cleanAborted will clog the executor service, we should do something about this, either in this patch, or follow up something like https://issues.apache.org/jira/browse/HIVE-21150 immediately after this.

deniskuzZ replied:
In original patch Map<String, NonReentrantReadWriteLock> tableLock = new ConcurrentHashMap<>() was used to prevent  a concurrent p-clean (where the whole table will be scanned). I think, that is resolved by grouping p-cleans and recording list of writeIds that needs to be removed. @vpnvishv is that correct?

pvargacl replied:
I agree, it can be addressed in a follow up Jira.

pvargacl commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorMR.java:
```diff
@@ -237,6 +237,7 @@ void run(HiveConf conf, String jobName, Table t, Partition p, StorageDescriptor
     }
 
     JobConf job = createBaseJobConf(conf, jobName, t, sd, writeIds, ci);
+    QueryCompactor.Util.removeAbortedDirsForAcidTable(conf, dir);
```
@vpnvishv Why do we do this here? I understand we can, but why don't we let the Cleaner to delete the files? This just makes the compactor slower. Do we have a functionality reason for this?
After this change it will run in CompactorMR and in MMQueryCompactors, but not in normal QueryCompactors?

vpnvishv replied:
@pvargacl You are right, this is not required, as now compactor run in a transaction and the cleaner has validTxnList with aborted bits set. This we have added wrt to Hive-3, in which cleaner doesn't have aborted bits set, as we create validWriteIdList for cleaner based on the highestWriteId.

deniskuzZ replied:
removed

pvargacl commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:
```diff
@@ -107,11 +107,12 @@ public CompactionTxnHandler() {
-            + "MIN(\"TXN_STARTED\"), COUNT(*)"
+        String sCheckAborted = "SELECT \"TC_DATABASE\", \"TC_TABLE\", \"TC_PARTITION\", "
+            + "MIN(\"TXN_STARTED\"), COUNT(*), "
+            + "MAX(CASE WHEN \"TC_OPERATION_TYPE\" = " + OperationType.DYNPART + " THEN 1 ELSE 0 END) AS \"IS_DP\" "
```
I might be mistaken here, but does this mean, that if we have many "normal" aborted txn and 1 aborted dynpart txn, we will not initiate a normal compaction until the dynpart stuff is not cleaned up? Is this ok, shouldn't we doing both?

deniskuzZ replied:
why is that? aborted dynPart is just a special case that would be handled separately (IS_DP=1).

pvargacl replied:
Previously if you had aborted txn above threshold this would generate a "normal" compaction that would clean up everything. However now if you have one dynpart aborted the type will be CLEAN_ABORTED that will only clean the writeids belonging to p-type records and leave everything else. This will delay the normal cleaning. I am not sure that is a problem or not.

deniskuzZ replied:
I still don't follow. Aborted txn check is done per db/table/partition, so if you have db1/tbl1/{p1-p100}/type=NOT_DP and db1/tbl1/null/type=DP - that should generate 2 entries in potential compactions.

deniskuzZ replied:
oh, sorry, I only considered time based threshold

deniskuzZ commented:
> @deniskuzZ Overall change LGTM.
> Looked into the test failures, one of test requires change in expected values wrt master branch. Other two looks genuine failures to me. Please check the inline comments.

@vpnvishv, thank you for the review!

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:
```diff
@@ -2839,6 +2848,87 @@ public static void setNonTransactional(Map<String, String> tblProps) {
+    PathFilter filter = (p) -> {
+      String name = p.getName();
+      for (Long wId : writeIds) {
+        if (name.startsWith(deltaSubdir(wId, wId)) && !name.contains("=")) {
```
Why would it read the aborted data as valid if txn is in aborted state?

vpnvishv replied:
@pvargacl Sorry I may be missing something here, but with this change, how can compactor read the data of an aborted delta. It should be in the aborted list right, due to this dummy p type entry in TXN_COMPONENTS?

pvargacl replied:
You are right, I got confused, the p entry will solve this.

deniskuzZ commented:
@vpnvishv I can't find any tests that would actually cover dynamic part abort. HIVETESTMODEFAILLOADDYNAMICPARTITION is not used anywhere. Am I missing something here?

vpnvishv commented:
@deniskuzZ Yes.. Sorry I missed those while porting changes from my private branch to open source Hive-3.
I think you have already covered it now. Still if you want to refer I have Updated Hive-3 PR with those tests, https://github.com/apache/hive/pull/1415/commits/4596b50e0f4e00c363a744bad13a8c750330d9a0

deniskuzZ commented:
> @deniskuzZ Yes.. Sorry I missed those while porting changes from my private branch to open source [Hive-3](https://issues.apache.org/jira/browse/HIVE-3).
> I think you have already covered it now. Still if you want to refer I have Updated [Hive-3](https://issues.apache.org/jira/browse/HIVE-3) PR with those tests, [4596b50](https://github.com/apache/hive/commit/4596b50e0f4e00c363a744bad13a8c750330d9a0)

Thanks @vpnvishv, I'll check them. Also noticed that new test property is not actually configured. Fixed in master patch.

deniskuzZ commented:
looks like master is broken right now

pvargacl commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java:
```diff
@@ -853,6 +857,273 @@ public void majorCompactAfterAbort() throws Exception {
+    RemoteIterator it =
+        fs.listFiles(new Path(table.getSd().getLocation()), true);
+    if (it.hasNext()) {
+      Assert.fail("Expecting compaction to have cleaned the directories, FileStatus[] stat " + Arrays.toString(stat));
```
I think this assert is quit misleading. I might be wrong but the recursive listing skips empty directories, and actually this new version of cleaning will keep the partition directories (it should) and only delete the delta dirs and files.

deniskuzZ replied:
yes, recursive listing skips empty directories - that's was done intentionally. changed the comment

pvargacl commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java:
```diff
@@ -853,6 +857,273 @@ public void majorCompactAfterAbort() throws Exception {
+    connection.write("2,2".getBytes());
+
+    int count = TxnDbUtil.countQueryAgent(conf, "select count(*) from TXN_COMPONENTS where TC_OPERATION_TYPE='p'");
+    // We should have two rows corresponding to the two aborted transactions
```
this comment is copied I guess

deniskuzZ replied:
fixed

pvargacl commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java:
```diff
@@ -853,6 +857,273 @@ public void majorCompactAfterAbort() throws Exception {
+
+  }
+
+  private HiveStreamingConnection prepareTableAndConnection(String dbName, String tblName, int batchSize) throws Exception {
```
Could you add two more test with batch size 2. first batch writes to p1 and p2 and commits, second batch writes to p2 and p3 and aborts. And one with the aborted / committed order changed.

deniskuzZ replied:
added

pvargacl commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:
```diff
@@ -232,6 +240,51 @@ public Object run() throws Exception {
+
+  private void cleanAborted(CompactionInfo ci) throws MetaException {
+    if (ci.writeIds == null || ci.writeIds.size() == 0) {
+      LOG.warn("Attempted cleaning aborted transaction with empty writeId list");
```
Shouldn't you mark the compaction failed or cleaned?

deniskuzZ replied:
yep, good catch!

deniskuzZ replied:
fixed

pvargacl commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:
```diff
@@ -414,77 +436,56 @@ public void markCleaned(CompactionInfo info) throws MetaException {
-        if(info.highestWriteId != 0) {
-          pStmt.setLong(paramCount++, info.highestWriteId);
+        s = "DELETE FROM \"TXN_COMPONENTS\" WHERE \"TC_TXNID\" IN (" +
+          "   SELECT \"TXN_ID\" FROM \"TXNS\" WHERE \"TXN_STATE\" = " + TxnStatus.ABORTED + ") " +
```
cool

klcopp commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:
```diff
@@ -414,76 +403,30 @@ public void markCleaned(CompactionInfo info) throws MetaException {
-        if (info.highestWriteId != 0) s += " AND \"TC_WRITEID\" <= ?";
-        if (info.partName != null) s += " AND \"TC_PARTITION\" = ?";
-
+        s = "DELETE FROM \"TXN_COMPONENTS\" WHERE \"TC_TXNID\" IN (" +
```
This is just refactoring right? LGTM but can you make sure @pvary sees this as well?

deniskuzZ replied:
this is an optimization that makes everything in 1 db request instead of 2 (select + delete)

deniskuzZ replied:
@pvary, could you please take a quick look? thanks!

klcopp replied:
never mind, LGTM

deniskuzZ commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:
```diff
@@ -400,11 +389,11 @@ public void markCleaned(CompactionInfo info) throws MetaException {
         }
         if(info.highestWriteId != 0) {
-          pStmt.setLong(paramCount++, info.highestWriteId);
+          pStmt.setLong(paramCount, info.highestWriteId);
```
redundant post increment

deniskuzZ commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:
```diff
@@ -134,9 +132,6 @@ public CompactionTxnHandler() {
         }
-
-        LOG.debug("Going to rollback");
-        dbConn.rollback();
```
no idea :)

deniskuzZ commented on ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:
```diff
@@ -589,4 +593,9 @@ private void checkInterrupt() throws InterruptedException {
   }
-}
+
+  private static boolean isDynPartAbort(Table t, CompactionInfo ci) {
```
could be, do you know if there is some helper class I could move isDynPartAbort method?
